design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,480 @@
+Design and develop an API to handle live storage migration between PVCs.
+
+ ## Non  Goals
+*  This feature is not directly aimed at live migrating the VM using RWO volumes. Hence, the API covers and refers only to the volume migration.
+*  This design doesn't cover migration between storage classes, but rather implements basic APIs that can be used by overlying tooling.
+*  This design doesn't include the creation of the destination PVCs, it expects the destination PVCs to exist.
+
+ ## Definition of Users
+This feature is for all KubeVirt users who want to move the PVCs assigned to a VM without workload interruptions.
+
+## User Stories
+
+  1. As a VM owner, the storage class of volumes used by a running VM is deprecated, and the volumes need to be moved to another storage class.
+  2. As a VM owner, I want to use higher performance storage class that is now available in the cluster.
+  3. As a VM owner, I see my non-resizable storage device filling up and I want to move my VM to a larger storage device without interruptions in order to give it more storage.
+  4. As an infra admin, I need to replace the storage array due to a hardware refresh and I want to do this without stopping the VMs in order to make this infra operation transparent to VM owners.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD called `VolumeMigration`
+
+## Volume migration custom resource
+
+```golang
+type VolumeMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec VolumeMigrationSpec `json:"spec"valid:"required"`
+	Status VolumeMigrationStatus `json:"status,omitempty"`
+}
+
+type VolumeMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items []VolumeMigration `json:"items"`
+}
+
+type SourcePvcReclaimPolicy string
+const (
+	SourcePvcReclaimPolicyDelete SourcePvcReclaimPolicy = "Delete"
+	SourcePvcReclaimPolicyRetain SourcePvcReclaimPolicy = "Retain"
+)
+
+type MigratedVolume struct {
+	SourcePvc      string `json:"sourcePvc" valid:"required"`
+	DestinationPvc string `json:"destinationPvc" valid:"required"`
+	SourcePvcReclaimPolicy SourcePvcReclaimPolicy `json:"sourcePvcReclaimPolicy,omitempty"`
+}
+
+type VolumeMigrationSpec struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type VolumeMigrationPhase string
+
+const (
+	VolumeMigrationPhasePending    = "Pending"
+	VolumeMigrationPhaseRunning    = "Running"
+	VolumeMigrationPhaseScheduling = "Scheduling"
+	VolumeMigrationPhaseSucceeded  = "Succeeded"
+	VolumeMigrationPhaseFailed     = "Failed"
+	VolumeMigrationPhaseUnknown    = "Unknown"
+)
+
+type VolumeMigrationPhaseTransitionTimestamp struct {
+	Phase                    VolumeMigrationPhase `json:"phase,omitempty"`
+	PhaseTransitionTimestamp metav1.Time          `json:"phaseTransitionTimestamp,omitempty"`
+}
+
+type MigratedVolumeState string
```

Comment:
Confused by `MigratedVolumeState` vs `VolumeMigrationPhase` wondering if there are more descriptive names

Reply:
I'm not sure about the right names here. Suggestions are welcome! 

The `VolumeMigrationPhase` is similar to the [VirtualMachineInstanceMigrationPhase](VirtualMachineInstanceMigrationPhase) while `MigratedVolumeState` indicates the status of each migrated volume.

Reply:
I got the impression that `VolumeMigration` had two steps.  1) validate that the set of PVCs can in fact be migrated 2) do the migration.  I also got the impression that `MigratedVolumeState` does not change after step 1.  If both of those things are true maybe it makes sense to think as `MigratedVolumeState` a `VolumeValidation`?  Or maybe the Kubernetes condition model better fits somehow?

Reply:
Your understanding is correct. `MigratedVolumeState` don't change after the validation. If all the migrated volumes are valid only then we proceed with the migration. I thought to State because it could be pending, but I can change it to `VolumeValidation`

Reply:
I changed and renamed the field `MigratedVolumeState` to `VolumeValidation`